tests/docs/utils.py

```diff
@@ -2,13 +2,12 @@
 
 import re
 from pathlib import Path
-from typing import Union
+from typing import List, Union
 
 
-def extract_code_blocks(path: Union[str, Path]) -> str:
-    """Extract all python code blocks from the specified file into a single string."""
+def extract_code_blocks(path: Union[str, Path]) -> List[str]:
```

Comment:
This code only extracts and thus tests code that is written via triple backticks. As we discussed in our chat, we want to use triple tilde symbols when writing python pseudocode.
This implies that python pseudocode will not be checked for formatting, and can thus lead to the python pseudocode and the python code not being formatted in the same way without us being able to fix that automatically, unless we write a dedicated function for formatting the pseudocode.
My questions here are:
1.  Would such a function even work? That is, how does ruff behave for non-proper python code?
2. Does this require us to have guidelines on how to write pseudocode?
3. Would it be fine if the pseudocode is not formatted properly?

Reply:
Can we first clarify what exactly you and Martin mean when you speak about "pseudocode" :smile:? Are we talking about something like

```
run algorithm
write output to X
read variable Y
```

Or what exactly do you have in mind? Because for the above, I would not now how we could even expect to have a "consistent" formatting across cases!?

Reply:
I think @Scienfitz means something like this:
~~~python
campaign = Campaign(
    searchspace = some_searchspace,
    objective = some_objective,
    ... # additional optional stuff that is not important here
)
~~~
So code that still looks and feels "python" but is not proper python code since e.g. objects do not exist, notation and syntax is not entirely correct to improve readability, stuff is not imported and so on. That is at least my understanding of the kind of Pseudocode Martin would like.
See here a specific request from the campaign user guide discussions:
![image](https://github.com/emdgroup/baybe/assets/84016733/df25d17d-2ac6-4cfd-9f23-95b17fd5dae5)

Reply:
Ok, but then the situation is clear, no?
* Martins example is valid python syntax (although objects are not defined, but that doesn't matter for a code formatter). Therefore, it can and should be formatted. So in this case, we should use triple backticks instead of tilde (or is there another reason for the tilde?)
* If we really have an example that is NOT python syntax and should NOT formatted, we simply opt out. First, I would have said by using tilde instead of ticks, but now I wonder why you would add the the `python` keyword in the first place? If your pseudocode does NOT follow python syntax, there is also no point of adding the indentifier. So is there even a use case for having the tilde approach in the first place?

Reply:
The tilde is so that the docs test (i.e., the code that just concatenates everything and then executes it) skips that part of the code, so this is necessary and a minimal way to distinguish "This code should be tested" and "This code should not be tested".

For me, it would be fine if we agree to some statement "Pseudocode means that we use python syntax and a python code block enclosed by triple tilde. Pseudocode means that the code could be run if all necessary imports were made and mentioned objects were created." If we agree to this "definition" of pseudocode, then it should of course also be formatted accordingly.

With all of our discussions around writing the user guides, it is just important that we all agree on something like that. Would my proposed definition more or less work for all of us?

Reply:
seems this is still pre-dev-meeting discussion, I'm summarizing:
* triple ticks for python code that is also executable, this can be both format and execution checked
* triple tile for pseudo code, that should also be format checked but not execution checked

Reply:
Is now included 👍🏼

Reply:
Please double-check and resolve if OK